### DISCUSSION TITLE
Stitching together audio from websocket stream

### DISCUSSION THREAD

**Discussion by terato2** [savonet/liquidsoap#2858] [User Support]
Hello there!
I'm having trouble navigating the documentation and everything else (apart from liquidsoap) I tried has not been very helpful.
I have a websocket server sending me the current song, where to download it and when it should start, and the song that comes after it.

What I want to do is turn that into a single vorbis stream.
I suppose this is somehow possible by making an external program that transforms the websocket data into data manageable by liquidsoap and then using `request.dynamic`?
If there is a better way than using liquidsoap I'd be happy to hear that too! I wouldn't mind having to program a bit either as long as I at least know what to do.

I've though about somehow using ffmpeg but there doesn't seem to be any way to do what I want (that is, dynamically changing the input source).
Then I thought that Icecast could make this work but none of the libshout bindings are in languages I like.

Edit: I've just had another thought that might(?) work, would it be possible to make a FIFO pipe and then point ffmpeg to that in order to dynamically change  the input?

Another edit: I just figured out that liquidsoup can have a websocket input, after deciphering the source code I managed to figure that out (it's not in the docs seemingly?). Only problem is that this way I have no way of limiting the data that is sent to the websocket thus resulting in constant buffer overruns. Not sure whether or not this is a step in the right direction.

**Comment by Russsgithub** [comment]
@terato2 

request.queue gives you a source you can queue that you can push file uri's into over telnet or http(s) ( if you register the http(s) server ) . You can annotate cue in and out etc if you need that type of control inside liquidsoap.

https://www.liquidsoap.info/doc-dev/reference.html#request.queue

I'm guessing there is help in the book on this too ?

**Reply by terato2** [reply]
I was actually just looking at the possibility of this working and it does!!
I'm not entirely certain on whether or not I like this solution since it seems very hacky to me (with telnet I mean).
I didn't even know the book was available for free, definitely gonna take a look at it!

**Comment by Russsgithub** [comment]
@terato2 , telnet is one method of interacting with it.

You can interact over http to push information in, there's something at the bottom of this page

https://www.liquidsoap.info/doc-2.1.3/harbor_http.html

In get metadata section. Were a get request is registered. If you register a put request and set the function to take the data and push requests into the queue that'd work too. There will be more info in the docs and book about it.

I'm using my phone at the mo so hunting around for the relevant part of the docs is tricky, but both the docs and book are very good so you should find what you need.

The book is available as a pdf through the book link. Below the pic is a link.

Good luck.

**Comment by Russsgithub** [comment]
@terato2 

This is from the book, page 158

```
default = playlist("~/Music")
queue = request.queue()
radio = fallback(track_sensitive=false, [queue, default])
def play(~protocol, ~headers, ~data, uri)
log.important("Serving uri: #{uri}")
let (uri, args) = url.split(uri)
fname = args["file"]
title = args["title"]
if file.exists(fname) then
queue.push.uri("annotate:title=#{string.quote(title)}:#{fname}")
http.response(data="Request pushed.")
else
http.response(code=404, data="Invalid file.")
end
end
harbor.http.register("/play", play)
```
